tests/ga/test_extension.py

```diff
@@ -2535,304 +2536,285 @@ def test_ext_handler_with_systemd_cgroup_enabled(self, *args):
 
-        handler_json = {
-            "installCommand": "sample.py -install",
-            "uninstallCommand": "sample.py -uninstall",
-            "updateCommand": "sample.py -update",
-            "enableCommand": "sample.py -enable",
-            "disableCommand": "sample.py -disable",
-            "rebootAfterInstall": False,
-            "reportHeartbeat": False,
-            "continueOnUpdateFailure": continue_on_update_failure
-        }
+            # Run the handler with only the first version to install it
+            exthandlers_handler.run()
+            
+            # Verify the extension was picked up correctly.
+            # NOTE: this won't work for tests that want install to fail.
+            firstVersionActor.get_command("installCommand").assert_called_once()
+            firstVersionActor.get_command("enableCommand").assert_called_once()
 
-        if handler:
-            handler_json.update(handler)
+            # Update to the second goal state incarnation, including the updated actor this time
+            add_extension_actors(protocol, 2, secondVersionActor)
 
-        ext_handler_properties = ExtHandlerProperties()
-        ext_handler_properties.version = version
-        ext_handler = ExtHandler(name=name)
-        ext_handler.properties = ext_handler_properties
-        ext_handler_i = ExtHandlerInstance(ext_handler=ext_handler, protocol=None)
-        ext_handler_i.load_manifest = MagicMock(return_value=HandlerManifest({'handlerManifest': handler_json}))
-        fileutil.mkdir(ext_handler_i.get_base_dir())
-        return ext_handler_i
+            # Run the handler a second time on the updated goal state (including the new extension version)
+            exthandlers_handler.run()
+    
+    
+    def test_enabled_ext_should_be_disabled_at_ver_update(self):
 
-    def test_disable_failed_env_variable_should_be_set_for_update_cmd_when_continue_on_update_failure_is_true(
-            self, *args):
-        old_handler_i = self._get_ext_handler_instance('foo', '1.0.0')
-        new_handler_i = self._get_ext_handler_instance('foo', '1.0.1', continue_on_update_failure=True)
+        first_ext = get_extension_actor()
+        second_ext = get_extension_actor(version="1.1.0")
 
-        with patch.object(CGroupConfigurator.get_instance(), "start_extension_command",
-                          side_effect=ExtensionError('disable Failed')) as patch_start_cmd:
-            with self.assertRaises(ExtensionError):
-                ExtHandlersHandler._update_extension_handler_and_return_if_failed(old_handler_i, new_handler_i)
+        TestExtensionUpdateOnFailure._do_upgrade_scenario(first_ext, second_ext)
 
-            args, kwargs = patch_start_cmd.call_args
+        first_ext.get_command("disableCommand").assert_called_once()
+        second_ext.get_command("installCommand").assert_called_once()
+        second_ext.get_command("enableCommand").assert_called_once()
```

Comment:
Do we have a test for the entire update sequence? if we do, then this test may be redundant. If we don't, can we turn this test into one that verifies the full sequence of extension commands? (and also assert that the commands are called in the right order instead of just called once)

Reply:
I do not believe there is a "nothing failed" test, at least in this class. It would be trivial to add checks for the entire sequence of commands, and I think I have an idea for guaranteeing ordering... I'll toy around with it and commit a change.

Reply:
I've definitely got the "guaranteed ordering" working. Do you think it would be worth it to integrate the functionality into extension_actor, enabled by default? I.e. is this something we'll want for all the tests?

Reply:
Not sure yet, let me make more progress in the review to get a feeling of what the actor is responsible for